upstream.go

```diff
@@ -0,0 +1,65 @@
+package ngrok
+
+import (
+	"crypto/tls"
+)
+
+// Upstream represents configuration for forwarding to an upstream service.
+type Upstream struct {
+	addr            string
+	protocol        string
+	proxyProto      string
+	tlsClientConfig *tls.Config
+	dialer          Dialer
+}
+
+// UpstreamOption configures an Upstream instance.
+type UpstreamOption func(*Upstream)
+
+// WithUpstream creates an Upstream configuration with a required address.
+// The address can be in various formats such as:
+// - "80" (a port number for local services)
+// - "example.com:8080" (a host:port combination)
+// - "http://example.com" (a full URL)
+func WithUpstream(addr string, opts ...UpstreamOption) *Upstream {
+	opt := &Upstream{addr: addr}
+	for _, o := range opts {
+		o(opt)
+	}
+	return opt
+}
+
+// WithUpstreamProtocol sets the protocol to use when forwarding to the upstream.
+// This is typically used to specify "http2" when communicating with an
+// upstream HTTP/2 server.
+func WithUpstreamProtocol(proto string) UpstreamOption {
+	return func(o *Upstream) {
+		o.protocol = proto
+	}
+}
+
+// WithUpstreamTLSClientConfig sets the TLS client configuration to use when connecting
+// to the upstream server over TLS.
+func WithUpstreamTLSClientConfig(config *tls.Config) UpstreamOption {
+	return func(o *Upstream) {
+		o.tlsClientConfig = config
+	}
+}
+
+// WithUpstreamProxyProto sets the PROXY protocol version to use when connecting
+// to the upstream server. Valid values are "v1" or "v2".
+//
+// See https://ngrok.com/docs/agent/config/v3/#upstreamproxy_protocol
+func WithUpstreamProxyProto(proxyProto string) UpstreamOption {
```

Comment:
Should we have an enum, like `ProxyProtocolV1` / `V2`, for this?

The string feels error-prone because both "1" and "v1" seem like reasonable inputs.

Reply:
resolved and fixed